msbuild/Xamarin.MacDev.Tasks.Core/Tasks/CreateBindingResourcePackageBase.cs

```diff
@@ -0,0 +1,97 @@
+		public string SessionId { get; set; }
+		
+		[Required]
+		public string OutputPath { get; set; }
+		
+		[Required]		
+		public ITaskItem[] NativeReferences { get; set; }
+		
+		[Required]
+		public string ProjectDir { get; set; }
+		
+		[Required]
+		public string BindingAssembly { get; set; }
+		
+		public override bool Execute ()
+		{
+			// LinkWith must be migrated for NoBindingEmbedding styled binding projects
+			if (NativeReferences.Length == 0)
+				Log.LogError ($"Unable to create a Binding Resource Package with no Native References. Move from LinkWith.cs to NativeReferences or disable NoBindingEmbedding");
+
+			string bindingResourcePath = Path.Combine (ProjectDir, OutputPath, $"{Path.GetFileNameWithoutExtension (BindingAssembly)}.resources");
+			Log.LogMessage ($"Creating binding resource package: {bindingResourcePath}");
+
+			Directory.CreateDirectory (bindingResourcePath);
+			foreach (var nativeRef in NativeReferences)
+				CopyNativeReference (nativeRef.ItemSpec, bindingResourcePath);
+
+			CreateManifest (bindingResourcePath);
+
+			return true;
+		}
+
+		// There is no built-in C# API for recursive copy of directories
+		void CopyNativeReference (string nativeRef, string resourcePath)
+		{
+			// TODO - Should this be smarter if we didn't change? Can't use smart copy unless framework? rsync?
+			// TODO - Is this safe for Windows builds?
+			var copyTask = new Microsoft.Build.Tasks.Exec {
+				Command = $"cp -r {nativeRef} {resourcePath}",
+				BuildEngine = this.BuildEngine
+			};
+			copyTask.Execute ();
+			if (copyTask.ExitCode != 0)
+				Log.LogError ($"Binding resource packaging copy exited with code {copyTask.ExitCode}");
+		}
+
+		string [] NativeReferenceAttributeNames = new string [] { "Kind", "ForceLoad", "SmartLink", "Frameworks", "WeakFrameworks", "LinkerFlags", "NeedsGccExceptionHandling", "IsCxx"};
+
+		void CreateManifest (string resourcePath)
+		{
+			XmlWriterSettings settings = new XmlWriterSettings() {
+				OmitXmlDeclaration = true,
+				Indent = true,
+				IndentChars = "\t",
+			};
+
+			using (var writer = XmlWriter.Create (Path.Combine (resourcePath, "manifest"), settings)) {
+				writer.WriteStartElement ("BindingAssembly");
+				writer.WriteAttributeString ("MVID", MVIDHelpers.FindAssemblyMVID (Path.Combine (ProjectDir, BindingAssembly)));
+
+				foreach (var nativeRef in NativeReferences) {
+					writer.WriteStartElement ("NativeReference");
+					writer.WriteAttributeString ("Name", Path.GetFileName (nativeRef.ItemSpec));
+
+					foreach (string attribute in NativeReferenceAttributeNames) {
+						writer.WriteStartElement (attribute);
+						writer.WriteString (nativeRef.GetMetadata (attribute));
+						writer.WriteEndElement ();
+					}
+
+					writer.WriteEndElement ();
+				}
+				writer.WriteEndElement ();
+			}
+		}
+	}
+
+	public static class MVIDHelpers {
+		public static string FindAssemblyMVID (string path)
+		{
+			var assembly = Assembly.ReflectionOnlyLoadFrom (path);
```

Comment:
that's looks costly - and also address something common ?
is anything else (new or existing) loading the assembly ?
Last week there were some guidelines mentioned for msbuild that included how to cache things (in case someone else already loads it)

Reply:
CopyTask copies files, not folders. I was concerned it wouldn't handle frameworks with symlinks correctly. I could test, or rework to use clone as Steve suggested.

This was added to catch the concern Manuel\Miguel raised where people could copy pasted a mismatch between the managed assembly and it's packaging. I don't think I can cache it.